modbus_init, Luxtronic2 reconnect and pack/unpack and smartvisu type error

kr2 opened the pull request:
modbus_init: [readme](https://github.com/kr2/smarthome/tree/develop/plugins/modbus)
Luxtronic2: [readme](https://github.com/kr2/smarthome/tree/develop/plugins/luxtronic2)

The count string handed from smartvisu was not casted to int.

kr2 committed:
fixed type error with smartvisu

kr2 committed:
added stuff

kr2 committed:
added pack unpack support per item

kr2 committed:
added reconnect support and some doc

kr2 committed:
added modbus_init

kr2 committed:
switched to threading instead of scheduler

kr2 committed:
switched to threading instead of scheduler

kr2 committed:
Merge branch 'develop-luxtronic' into develop

kr2 committed:
fixed type error

ohinckel commented on plugins/luxtronic2/README.md:
```diff
@@ -46,6 +56,12 @@ Defines a mapping to a attribute (read-only). All attribute values are bytes (nu
 
+### lux2_unpack
+Python lambda function. Called before the value is sent towards the device.
+
```
The unpack option sounds like the common `eval` setting for items. It evaluates the value and puts the result into the item. IMHO, I prefer to use common settings instead of having special plugin settings for all the plugins.

ohinckel commented on plugins/luxtronic2/README.md:
```diff
@@ -46,6 +56,12 @@ Defines a mapping to a attribute (read-only). All attribute values are bytes (nu
+
+### lux2_pack
+Called before the value is written to the smarthome item.
+
```
This could also be a common setting for items - something like `eval_out` - which works like `eval`, but the other way around when item is set.

kr2 replied:
I see your point and I thought about it too. Something like eval_out would work most of the time, but it would fail for something like the example below. My thinking is that a plugin should provide you with a ready to use value and the other way around should take the unencoded value. E.g. for the knx plugin you have to provide knx_dpt and the plugin handles pack/unpack stuff. 
First I only created *_pack for something like below and the way I think of plugins and I did *_unpack out of symmetry. E.g. one could use *_unpack to do the word/byte/... decoding stuff and eval for some value transformation (seconds to days).
I can make the *_pack and *_unpack optional, populated to hand the values through, but I would like to keep them.


```
 [[[dp1]]]
      type = num
      knx_dpt = 9
      knx_send = 10/4/8
      modbus_addr = Blub1 | 1 | 9 | 1
      modbus_type = HoldingRegister
      modbus_readInterval = -1
      modbus_pack = lambda x: [int((x * 1023) / 102.3)]
```

kr2 replied:
@ohinckel I looked a bit further into default values for modbus_pack (same for unpack and lux2). 
I still don't see a good way for `eval_out`. To make something like my last example work, it would have to check the caller e.g. like: `eval_out = return knxVal if caller == 'KNX' else return modbusVal`
I gonna stick with pack(/unpack), if you don't have a better solution.
The somewhat defaults at the moment are `lambda x: x`. I think this means one does not have to use the pack/unpack functions if the type is set to `list`. If *_pack is used it could be any other type and it should be possible to use eval for unpacking. I did not test that.
So again I don't see any better way.

ohinckel commented on plugins/luxtronic2/__init__.py:
```diff
@@ -226,51 +247,101 @@ def refresh_calculated(self):
+        # fail in this application, because it calles the foo without checking
+        # if the last call finished. This could clog up the system.
+        threading.Thread(name='Lux2_cycle', target=self.__refresh).start()
+        threading.Thread(name='Lux2_recon', target=self._reconnect).start()
```
Are there really cases were the update cylce needs too much time? Maybe using a short timeout for querying the device could help here.

kr2 replied:
See also my [bug report](https://github.com/mknx/smarthome/issues/208) for that. For now, it works with threading, but I could not get it working with the scheduler. I don't think it is about the socket timeout, because I also had problems with the scheduler in the modbus plugin too.

ohinckel commented on plugins/luxtronic2/__init__.py:
```diff
@@ -328,22 +399,44 @@ def _decode(self, identifier, value):
     def parse_item(self, item):
+        def __reverseListOp(param):
+            # reversing list building of smarthome.py config parser for the
+            # binary or opperator |
```
I think putting the value in quotes will fix this problem too. Usually you should be using quotes around the value to avoid escaping problem.

Other setting may have this problem too, but we should not start to implement such an "unquoting mechanism" for each of these settings. A general solution, which already exists (using quotes), would be more clear and consistent.

kr2 replied:
I did not know that quotes are supported, but if it works it's most definitely a better way to do it. I am going to fix it when there is time for it.

kr2 replied:
@ohinckel I just check the config parser again and quotes are not supported right now. I also tried it and there is no magic gunk happening somewhere else.

config.py line:89...
```python
attr, __, value = line.partition('=')
...
attr = attr.strip()
...
if '|' in value:
    item[attr] = [strip_quotes(x) for x in value.split('|')]
else:
    item[attr] = strip_quotes(value)
```

So to allow list operators in combination with quotes, it would have to ignore pipes inside of quotes and split if there are pips outside of them. I did a quick hack, see below. If you have a better solution or if I missed something let me know.

```python
def split_pipe(s):
    output = []
    split_pos = []
    quotes = ['"',"'"]
    pipe = '|'
    in_quoate = False
    start_quote = None
    
    for i in range(0, len(s)):
        if not in_quoate and s[i] == pipe:
            split_pos.append(i)
        
        if not in_quoate and s[i] in quotes:
            start_quote = s[i]
            in_quoate = True  
        elif in_quoate and s[i] == start_quote:
            in_quoate = False
            
    if len(split_pos) > 0:      
        output.append(s[0:split_pos[0]])
        for i in range(0,len(split_pos)):
            if i+1 < len(split_pos):
                output.append(s[split_pos[i]+1:split_pos[i+1]])
            else:
                output.append(s[split_pos[i]+1:])
        output = [strip_quotes(x) for x in output]
        return output
    else:
        return strip_quotes(s)

s = """'123|456'|"789'10'|1112"|131415 """

current = [strip_quotes(x) for x in s.split('|')]
for i in current:
    print(i)
print('-'*10)
x = split_pipe(s)
for i in x:
    print(i)
print('-'*10)
x = split_pipe('1')
print(x)
```

kr2 committed:
modbus added error handling for readloop and timertick

kr2 committed:
Added retries if bad telegrams occure e.g. crc before reconnect
Added some debug gunk. Verbose output option, status for each master in var/modbus/*.txt, ...
Added random start time for reads to even out load

kr2 committed:
sqlite fixed missing check for 'init'

kr2 commented:
Commit 	8e29643  will fix #209
